Add flag to enable prometheus metrics

qezz opened the pull request:
#### Problem

Add flag to expose prometheus metrics.
The metrics shouldn't be exposed by default.

#### Summary of Changes

Add condition flag on `/metrics` endpoint

#### Possible improvements

- naming could be better
- add a way to specify the metrics endpoint e.g. 
   - convert the flag into an option`--rpc-prometheus-metrics "/metrics"`
   - or add another option `--rpc-prometheus-metrics --rpc-prometheus-metrics-endpoint "/metrics"`
   - **but** if we want to add a way to specify an endpoint, we have to deal with possible conflicts. So I would not add it.

qezz commented:
Still have to test it on a real instance

ruuda commented on core/src/validator.rs:
```diff
@@ -809,6 +810,7 @@ impl Validator {
                 max_complete_transaction_status_slot,
                 prioritization_fee_cache.clone(),
                 config.vote_accounts_to_monitor.clone(),
+                rpc_enable_prometheus_metrics,
```
Appending the new field at the end is the obvious thing to do, but upstream Solana also does that, so it tends to create conflicts when we have to rebase the patches. If you can insert it somewhere in the middle, it makes it easier to keep the patches up to date.

qezz replied:
Yeah, it's a bit of a mess already

> it tends to create conflicts when we have to rebase the patches

Ah, totally makes sense

qezz replied:
I moved it a bit up, but not everywhere.

ruuda commented on rpc/src/rpc.rs:
```diff
@@ -161,6 +161,7 @@ pub struct JsonRpcConfig {
     pub obsolete_v1_7_api: bool,
     pub rpc_scan_and_fix_roots: bool,
     pub max_request_body_size: Option<usize>,
+    pub rpc_prometheus_metrics: bool,
```
Maybe a bit more obvious:

```suggestion
    pub rpc_enable_prometheus_metrics: bool,
```

(I see it’s already called `_enable_` in other places, I think it makes sense to use it here too.)

ruuda commented on rpc/src/rpc_service.rs:
```diff
@@ -84,6 +84,7 @@ struct RpcRequestMiddleware {
     /// Initialized based on vote_accounts_to_monitor, maps identity
     /// pubkey associated with the vote account to the validator info.
     identity_info_map: Arc<IdentityInfoMap>,
+    enable_prometheus_metrics: Arc<AtomicBool>,
```
```suggestion
    enable_prometheus_metrics: bool,
```

I think we can get away with a regular bool here, the value does not change at runtime, and unlike e.g. `vote_accounts_to_monitor` it is small, so it makes sense to just copy the bool so every thread has its own bool. (For `vote_accounts_to_monitor`, the value also does not change at runtime, but the hash map is larger than a few pointers, so we can share a single instance among all threads by putting it in an `Arc`. But the bool is a single byte, the `Arc` is much larger than the bool itself.)

qezz replied:
done

ruuda commented on validator/src/main.rs:
```diff
@@ -1543,6 +1543,12 @@ pub fn main() {
                 .help("The maximum request body size accepted by rpc service"),
         )
+        .arg(
+            Arg::with_name("rpc_prometheus_metrics")
+                .long("rpc-prometheus-metrics")
```
```suggestion
            Arg::with_name("enable_rpc_prometheus_metrics")
                .long("enable-rpc-prometheus-metrics")
```
Not sure if there is already this inconsistency, but I see one arg that starts with `enable` below, it makes sense to me for boolean flags.

About `rpc` in the name, I’m not sure, on the one hand `--enable-rpc-prometheus-metrics` sounds like it would enable Prometheus metrics about RPC calls, which is not what the feature does. But on the other hand, it does listen on the RPC endpoint :thinking: I think just `--enable-prometheus-metrics` would be slightly clearer, but I don’t have a strong opinion on it.

qezz replied:
Yeah... if we don't want to tie it to the `rpc` name, it's better to omit this prefix.
IMO `--enable-prometheus-metrics` sounds better.

Anyway, one of the reasons I've added the `rpc` prefix is because its value goes to the Rpc service config, and also because the endpoint is on the RPC port.

qezz replied:
changed to `enable-prometheus-metrics`

ruuda reviewed (commented):
Nice, thanks! I have a few small comments.

qezz commented:
It needs to be squashed

ruuda commented on rpc/src/rpc_service.rs:
```diff
@@ -737,6 +743,7 @@ mod tests {
             RpcHealth::stub(),
             block_commitment_cache,
             None,
+            false,
```
There are some compile errors here because the argument moved, you can typecheck the tests with `cargo check --test`.

qezz replied:
@ruuda I wonder if this issue is on the `prometheus-master` branch as well 
e.g. https://github.com/solana-labs/solana/pull/27665/files#diff-ae00f1b802baaa57eb71ae6bf8425ca1c4e8e32b3e82ef2f5a122ac2307cb01dR691

ruuda replied:
Hmm, might be ... could you address this nonetheless?

We also have a `prometheus-1.13` branch that is compatible with Solana 1.13. It is more up to date because that is the one we actually use, maybe it is fixed there, but it doesn’t target the `master` branch.

qezz replied:
yes, sure sure

ruuda reviewed (commented):
There are still compile errors in the tests. Other than that, this looks good!

qezz committed:
Add flag to enable prometheus metrics

qezz committed:
Fix rpc service tests

qezz committed:
Fix replica-node tests

qezz committed:
Add missing with_suffix method to Metric

qezz commented:
Rebased onto `prometheus-1.13`.
Fixed the tests related to wrong args.

Also, added `with_suffix` to `Metric` because it failed to build tests otherwise - https://github.com/ChorusOne/solana/pull/75/commits/d0a0f11229de0fa32013528b8a45404c05964cc4
I actually don't know how did it work without this. Feel free to cherry pick

ruuda commented on prometheus/src/utils.rs:
```diff
@@ -94,6 +94,11 @@ impl<'a> Metric<'a> {
+    pub fn with_suffix(mut self, suffix: &'a str) -> Metric<'a> {
+        self.suffix = suffix;
+        self
+    }
```
Why is this change in here? It doesn’t appear to be used.

qezz replied:
It is used here (branch `prometheus-1.13`, `prometheus/src/utils.rs`)

https://github.com/ChorusOne/solana/blob/0dfa233d88fa38184339929d589db99f40a9568d/prometheus/src/utils.rs#L189-L199

Compilation errors otherwise

qezz replied:
Sorry, to clarify, errors when running `cargo check --tests`

```
error[E0599]: no method named `with_suffix` found for struct `utils::Metric` in the current scope
   --> prometheus/src/utils.rs:195:26
    |
49  | pub struct Metric<'a> {
    | --------------------- method `with_suffix` not found for this struct
...
195 |                         .with_suffix("_bucket")
    |                          ^^^^^^^^^^^ method not found in `utils::Metric<'_>`
```

ruuda replied:
Ah, ok, :+1: then, ignore my comment!

ruuda reviewed (approved):
LGTM with one nit!
